lib/Jsrt/Jsrt.cpp

```diff
@@ -156,6 +204,198 @@ CHAKRA_API JsCreateRuntime(_In_ JsRuntimeAttributes attributes, _In_opt_ JsThrea
     });
 }
 
+//A create context function that we can funnel to for regular and record or debug aware creation
+JsErrorCode CreateContextCore(_In_ JsRuntimeHandle runtimeHandle, _In_ bool createUnderTT, _Out_ JsContextRef *newContext)
+{
+    return GlobalAPIWrapper([&]() -> JsErrorCode {
+        PARAM_NOT_NULL(newContext);
+        VALIDATE_INCOMING_RUNTIME_HANDLE(runtimeHandle);
+
+        JsrtRuntime * runtime = JsrtRuntime::FromHandle(runtimeHandle);
+        ThreadContext * threadContext = runtime->GetThreadContext();
+
+        if(threadContext->GetRecycler() && threadContext->GetRecycler()->IsHeapEnumInProgress())
+        {
+            return JsErrorHeapEnumInProgress;
+        }
+        else if(threadContext->IsInThreadServiceCallback())
+        {
+            return JsErrorInThreadServiceCallback;
+        }
+
+        ThreadContextScope scope(threadContext);
+
+        if(!scope.IsValid())
+        {
+            return JsErrorWrongThread;
+        }
+
+#if ENABLE_TTD
+        if(createUnderTT & !threadContext->IsTTRequested)
+        {
+            AssertMsg(false, "Can't create a context under TT if runtime is not set to support TT!!!");
+            return JsErrorCategoryUsage;
+        }
+
+        if(createUnderTT & (threadContext->IsTTRecordRequested & threadContext->IsTTDebugRequested))
+        {
+            AssertMsg(false, "Can't set both of these at the same time!!!");
+            return JsErrorCategoryUsage;
+        }
+#endif
+
+        JsrtContext * context = JsrtContext::New(runtime);
+
+#if ENABLE_TTD
+        if(createUnderTT && threadContext->IsTTRequested && !threadContext->IsTTDInitialized())
+        {
+#if ENABLE_TTD_DEBUGGING
+            threadContext->SetThreadContextFlag(ThreadContextFlagNoJIT);
+#endif
+
+            if(threadContext->TTDInitializeTTDUriFunction == nullptr)
+            {
+                AssertMsg(false, "Must set this callback function first!!!");
+                return JsErrorCategoryUsage;
+            }
+
+            wchar_t* ttdlogStr = nullptr;
+            threadContext->TTDInitializeTTDUriFunction(threadContext->TTDUri, &ttdlogStr);
+
+            threadContext->InitTimeTravel(ttdlogStr, threadContext->IsTTRecordRequested, threadContext->IsTTDebugRequested, threadContext->TTSnapInterval, threadContext->TTSnapHistoryLength);
+
+            CoTaskMemFree(ttdlogStr);
```

Comment:
Is it guaranteed that these memory is allocated by host using CoTaskMemAlloc? The JsTTDInitializeUriCallback callback doesn't mention that.